subt/docker/robotika/ros_proxy_node.cc

```diff
@@ -0,0 +1,531 @@
+  /// "/x1/cmd_vel" if _name is specified as "x1".
+  /// \param[in] _name Name of the robot.
+  public: Controller(const std::string &_name);
+
+  /// \brief A function that will be called every loop of the ros spin
+  /// cycle.
+  public: void Update();
+
+  /// \brief Callback function for message from other comm clients.
+  /// \param[in] _srcAddress The address of the robot who sent the packet.
+  /// \param[in] _dstAddress The address of the robot who received the packet.
+  /// \param[in] _dstPort The destination port.
+  /// \param[in] _data The contents the packet holds.
+  private: void CommClientCallback(const std::string &_srcAddress,
+                                   const std::string &_dstAddress,
+                                   const uint32_t _dstPort,
+                                   const std::string &_data);
+
+  /// \brief ROS node handler.
+  private: ros::NodeHandle n;
+
+  /// \brief publisher to send cmd_vel
+  public: ros::Publisher velPub;
+
+  /// \brief Communication client.
+  private: std::unique_ptr<subt::CommsClient> client;
+
+  /// \brief Client to request pose from origin.
+  ros::ServiceClient originClient;
+
+  /// \brief Service to request pose from origin.
+  subt_msgs::PoseFromArtifact originSrv;
+
+  ros::Subscriber subClock;
+  ros::Subscriber subImu;
+  ros::Subscriber subScan;
+  ros::Subscriber subImage;
+  ros::Subscriber subOdom;
+
+  /// \brief True if robot has arrived at destination.
+  public: bool arrived{false};
+
+  /// \brief True if started.
+  private: bool started{false};
+
+  /// \brief Last time a comms message to another robot was sent.
+  private: std::chrono::time_point<std::chrono::system_clock> lastMsgSentTime;
+
+  /// \brief Name of this robot.
+  private: std::string name;
+
+  private: double prev_dist2{0.0};
+
+  public: bool ReportArtifact(subt::msgs::Artifact& artifact)
+  {
+    std::string serializedData;
+    if (!artifact.SerializeToString(&serializedData))
+    {
+      ROS_ERROR("ReportArtifact(): Error serializing message [%s]",
+          artifact.DebugString().c_str());
+      return false;
+    }
+    else if (!this->client->SendTo(serializedData, subt::kBaseStationName))
+    {
+      ROS_ERROR("CommsClient failed to Send serialized data.");
+      return false;
+    }
+    return true;
+  }
+
+  public: bool getSpeedCmd(geometry_msgs::Twist& msg);
+  public: bool parseArtf(char *input_str, subt::msgs::Artifact& artifact);
+};
+
+/////////////////////////////////////////////////
+Controller::Controller(const std::string &_name)
+{
+  ROS_INFO("Waiting for /clock, /subt/start, and /subt/pose_from_artifact");
+
+  ros::topic::waitForMessage<rosgraph_msgs::Clock>("/clock", this->n);
+
+  // Wait for the start service to be ready.
+  ros::service::waitForService("/subt/start", -1);
+  ros::service::waitForService("/subt/pose_from_artifact_origin", -1);
+  this->name = _name;
+  ROS_INFO("Using robot name[%s]\n", this->name.c_str());
```

Comment:
It has been suggested to wait also for some robot specific message to be sure it is already loaded in the simulation.